sharedbuffers/mapped_struct.py

```diff
@@ -341,6 +342,75 @@ def unpack_from(cls, buf, offs, idmap = None):
         key, values = mapped_list.unpack_from(buf, offs, idmap)
         return cls(zip(key, values))
 
+class proxied_buffer(object):
+
+    @classmethod
+    def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0, header_packer=struct.Struct('L')):
+        header_packer.pack_into(buf, offs, len(obj))
+        offs += header_packer.size
+
+        end_offs = offs + len(obj)
+        buf[offs:end_offs] = obj
+
+        return end_offs
+
+    @classmethod
+    def unpack_from(cls, buf, offs, idmap = None, header_packer=struct.Struct('L')):
+        size, = header_packer.unpack_from(buf, offs)
+        offs += header_packer.size
+
+        return buffer(buf, offs, size)
+
+class proxied_ndarray(object):
+
+    @staticmethod
+    def _make_dtype_params(dtype):
+        names = dtype.names
+        if names:
+            # It is a Structured array
+            fields = dtype.fields
+            return [
+                (k, proxied_ndarray._make_dtype_params(fields[k][0]))
+                for k in names
+            ]
+        else:
+            return dtype.str
+
+    @classmethod
+    @cython.locals(offs = cython.ulonglong, implicit_offs = cython.ulonglong, header_offs = cython.ulonglong)
+    def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0, header_packer=struct.Struct('LL')):
+        header_offs = offs
+        offs += header_packer.size
+
+        offs = mapped_tuple.pack_into(obj.shape, buf, offs)
+        dtype_offs = offs
+
+        dtype_params = proxied_ndarray._make_dtype_params(obj.dtype)
+        if isinstance(dtype_params, str):
+            dtype_params = [dtype_params]
+
+        offs = mapped_list.pack_into(dtype_params, buf, offs)
+        data_offs = offs
+
+        header_packer.pack_into(buf, header_offs, dtype_offs, data_offs)
```

Comment:
Offsets should be relative, not absolute. Otherwise you'll run into trouble if you're mapping this object from a different offset.